docs/allocations.md

```diff
@@ -0,0 +1,280 @@
+
+[ perf record: Captured and wrote 30.868 MB perf.data (3817 samples) ]
+```
+
+When `perf` tells you that it lost a number of chunks it means that it lost data. If `perf` lost data, you have a few options:
+
+- Reduce the amount of work your program is doing. For every allocation, `perf` will need to record a stack trace.
+- Reduce the maximum "stack dump" that `perf` records by changing the `--call-graph dwarf` parameter to for example `--call-graph dwarf,2048`. The default is to record a maximum of 4096 bytes which gives you pretty deep stacks, if you don't need that you can reduce the number. The tradeoff is that the flame graph may show you `[unknown]` stack frames which means that there are missing stack frames there. The unit is bytes.
+- You can raise the number of the `-m` parameter which is the size of the ring buffer that `perf` uses in memory (in multiples of `PAGE_SIZE`, usually that is 4kB)
+- You can give up nice call graphs and replace `--call-tree dwarf` with `--call-tree fp` (`fp` stands for frame pointer).
+
+### Flame graphs
+
+After a successful `perf record`, you can invoke the following command line to produce an SVG file with the flame graph
+
+```bash
+perf script | \
+    /FlameGraph/stackcollapse-perf.pl - | \
+    swift demangle --simplified | \
+    /FlameGraph/flamegraph.pl --countname allocations \
+        --width 2000 > out.svg
+```
+
+Let's expand a little on what the above command does:
+
+- It runs `perf script` which dumps the binary information that `perf record` recorded into a textual form.
+- Next, we invoke `stackcollapse-perf` on it which transforms the stacks that `perf script` outputs into the right format for Flame Graphs,
+- then we invoke `swift demangle --simplified` which will give us nice symbol names,
+- and lastly we create the Flame Graph itself
+
+After this command has run (which may run for a while), you should have an SVG file that you can open in your browser.
+
+For the above example program, please see an example flame graph below. Note how you can hover over the stack frames and get more information. To focus on a sub tree, you can click any stack frame too.
+
+Generally, in flame graphs, the X axis just means "count", it does not mean time. In other words, whether a stack appears on the left or the right is random and not determined when that stack was live (this is different in flame _charts_).
+
+Note that this flame graph is _not_ a CPU flame graph, 1 sample means 1 allocation here and not time spent on the CPU. Also be aware that stack frames that appear wide don't necessary allocate themselves, it means that they or something they call has allocated a lot. For example, `BaseSocketChannel.readable` is a very wide frame, and yet, it does not allocate itself. It however calls other functions (such as other parts of SwiftNIO and AsyncHTTPClient) that do then allocate a lot. It may take a little while to get familiar with flame graphs but there are great resources available online.
```

Comment:
```suggestion
Note that this flame graph is _not_ a CPU flame graph, 1 sample means 1 allocation here and not time spent on the CPU. Also be aware that stack frames that appear wide don't necessarily allocate directly, it means that they or something they call has allocated a lot. For example, `BaseSocketChannel.readable` is a very wide frame, and yet, it is not a function which allocates directly. However, it calls other functions (such as other parts of SwiftNIO and AsyncHTTPClient) that do allocate a lot. It may take a little while to get familiar with flame graphs but there are great resources available online.
```

Reply:
Maybe point to http://www.brendangregg.com/flamegraphs.html as a starting point for the available resources?